crypto_app/views.py

```diff
@@ -0,0 +1,54 @@
+from django.shortcuts import render
+from .models import Cryptocurrency, Transaction, Portfolio, MarketData
+
+
+def index(request):
+    return render(request, "crypto_app/index.html")
+
+def cryptocurrencies(request):
+    cryptocurrencies = Cryptocurrency.objects.all()
+    return render(request,
+                  "crypto_app/cryptocurrencies.html",
+                  {"cryptocurrencies": cryptocurrencies})
+
+
+def transactions(request):
+    transactions = Transaction.objects.select_related("user", "cryptocurrency").all()
+
+    return render(request,
+                  "crypto_app/transactions.html",
+                  {"transactions": transactions})
+
+
+def portfolio(request):
+    portfolio = Portfolio.objects.select_related("user", "cryptocurrency").all()
+
+    for portfolio_item in portfolio:
+        preferred_market_data = MarketData.objects.filter(
+            cryptocurrency=portfolio_item.cryptocurrency,
+            currency=portfolio_item.user.currency
+        ).order_by("-date").first()
+
+        if preferred_market_data:
+            portfolio_item.price = preferred_market_data.price
+            portfolio_item.currency = preferred_market_data.currency
+        else:
+            usd_market_data = MarketData.objects.filter(
+                cryptocurrency=portfolio_item.cryptocurrency,
+                currency="USD"
+            ).order_by("-date").first()
```

Comment:
Jeśli masz jakieś konkretne query, które będzie wykonywane często, to warto dodać indexy na bazie danych, aby pobieranie tych danych było szybsze. 
https://docs.djangoproject.com/en/5.2/ref/models/indexes/